### PR TITLE
Convert old sidebars to JSON and remove C++ sidebars.

### PR DESCRIPTION
**Pull request by Inglonias**
#### Summary
Interface "Hides the old sidebars and replaces them with fancy json sidebars"

#### Purpose of change
Implements the final near-term goal of #53957

#### Describe the solution
Using the code from #54191 as a baseline, I filled out the classic and compact sidebars. I am working under the assumption that the Label and Label Narrow sidebars are already correct.

#### Describe alternatives you've considered
Waiting for @wapcaplet to come back down from the mountains instead of work on this myself. If wapcaplet doesn't want me to continue working on this PR, that's fine by me. But as far as I can tell, their PR didn't stall out because of missing features.

The compact sidebar was made 4 characters wider to account for 2 pixel column padding. I thought about making column padding customizable, but... ehhhhhh...

#### Testing
I spent several hours of my June 12 evening dogfooding this change. All major remaining issues have been fixed.

#### Additional context
There are a few things I can't match one-to-one. ~~The big omission is that body status has been split out from health. This is clearer to new players, but I want to give players the option of having color-coding, which I will do in a separate PR.~~ #58603 resolves this. I also don't believe there is a way to see the mending status of a given limb anymore in the sidebar, though the medical UI offers that functionality. I plan to make further enhancements later to account for these issues, but they are out of scope here.

### CODE REVIEW

**Comment by Inglonias** [comment]
The C++ changes are to get past the "unused function" errors. I had better make damn sure that these sidebars are ready for primetime before I merge, huh?

**Comment by Inglonias** [comment]
> While I agree that this is something that needs to be done, we're currently missing support for "rows-in-columns" in custom widgets, which is needed to properly replace some of these layouts. I'd suggest holding off on removing the hardcoded sidebar options until then.

That didn't really stop me. I just used nested widgets for that. 

`  {
    "id": "lcla_health_layout",
    "type": "widget",
    "style": "layout",
    "label": "Health",
    "arrange": "columns",
    "widgets": [ "lcla_health_overmap", "lcla_health_left_layout", "lcla_health_mid_layout", "lcla_health_right_layout" ]
  },`

**Comment by Inglonias** [comment]
If there's enough call for the hardcoded sidebars to stay, I can just maintain a mod somewhere (probably not the main repo) for custom sidebars. No point throwing the work I did into the sea if I can help it.

**Comment by Inglonias** [comment]
So this is what I get for writing an entire PR without doing any testing first, huh?

It appears that what dseguin was referring to was this:
```
DEBUG    : widget layout called with rows

 FUNCTION : layout
 FILE     : D:\a\Cataclysm-DDA\Cataclysm-DDA\src\widget.cpp
 LINE     : 1382
 VERSION  : b70bca1
```

As far as I can tell, you can't have layout widgets with a row arrangement unless everything inside of it is a layout widget with column arrangements. Damn. Yeah, I can see why this was being held up. That's kind of disappointing, and I agree - This qualifies as a dealbreaker.

**Comment by Inglonias** [comment]
Reopening. New features make this much more feasible

**Comment by Inglonias** [comment]
Closing because every time I push, this is spamming me with emails. I will re-open this when it's ready.

**Comment by Inglonias** [comment]
Alright. I think this is close enough to ready to open up for testing.

**Comment by Inglonias** [comment]
Do not merge without merging #58355 first.

**Comment by snipercup** [comment]
I tested your pre-compiled version you shared in discord. This is the  default view after loading which I am not going to use but I'll just say that I was a bit confused about the temperature part where nothing is displayed behind the body parts. Maybe it should just default do 'comfortable'. Also the stamina bar is much more visible in this view then the others I share below:
image


I switched to compact and there's an empty body part status alt enabled by default in the compact view but I have no idea what it's supposed to show
image


The labels in the -env section should be highlighted to distinguish between label and value. Also the 'place' label wasn't there before and I think we can do without. Just show the place name and not the label and players can figure it out.
image


I don't know if it was like this before, but when you hide enough of the sidebar to lift the minimap up, it will still remain visible, as though it's duplicated. However the second one doesn't update anymore.
image
sidebars


Lastly I really had to look for the stamina bar, it's in the top-right. It's visible, but I prefer it be on the left like in the first screenshot
image

All in all I like what you do!

**Comment by Inglonias** [comment]
Body part status displays whether a body part has a bite, is bleeding, is infected, etc. Due to current limitations that information had to be separated from limb health

**Comment by snipercup** [comment]
I just noticed that the direction label doesn't light up on the compass even though he can see me:
image

image


It should light up like this (north):
image

**Comment by Inglonias** [comment]
@snipercup It is currently not possible to set the color of a label for a widget.

**Comment by snipercup** [comment]
> @snipercup It is currently not possible to set the color of a label for a widget.

Oh, I see. How will the player get the information that is visible in the current version of the game? I rely a lot on the compass highlight. Is there another way to display this information?

**Comment by Inglonias** [comment]
Would you believe that I didn't know the compass could do that after several years of playing this game?

I'll look into restoring that functionality this week.

**Comment by Inglonias** [comment]
So, I see a way to make the compass do the thing you describe. Trouble is that I don't know if it's a good idea to treat the compass special in the way I would need to make this work. I think in this case it's worth it. That said, I don't know if it's in scope for this PR. Since I was planning to explore the ability to color labels anyhow, the compass fix may go in that PR instead of this one.

**Comment by Inglonias** [comment]
So, I don't know what version of the game you are running, but the compass does not light up red when enemies can see me on the latest experimental. What actually causes it to light up is the danger level in that direction, which I feel is far less useful information. 
image

**Comment by snipercup** [comment]
No, the compass stops highlighting as soon as a zombie is in your view. You can tell from your screenshot, the south-east is the only direction where there are zombies that can see you but not in your view.

**Comment by Inglonias** [comment]
> No, the compass stops highlighting as soon as a zombie is in your view. You can tell from your screenshot, the south-east is the only direction where there are zombies that can see you but not in your view.

This is not correct, and [I have the code to prove it](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/display.cpp#L1463).

`nc_color c = unique_types[i].empty() && unique_mons[i].empty() ? c_dark_gray : ( dangerous[i] ? c_light_red : c_light_gray );`

**Comment by snipercup** [comment]
Well, if that's what the code says then that is what it does. All I care about is that the compass displays the same information after your change as it does now.

**Comment by Inglonias** [comment]
> Well, if that's what the code says then that is what it does. All I care about is that the compass displays the same information after your change as it does now.

I understand that and appreciate that view. My argument is that this is not a large enough regression to hold up this PR from being merged, though it does demand that someone (almost certainly me) go back in afterwards and restore the missing functionality. I am willing to do that, but feel that doing so is out of scope for this pull request. If people with merge permissions disagree, then obviously this PR doesn't get merged.

**Comment by Inglonias** [comment]
I have a version of the game compiled that restores the color coding of both limb health and the compass, though the compass is somewhat unreliable and I'm not sure why yet.

This work will be in a separate PR that should be merged alongside this one.
image

**Comment by Inglonias** [comment]
#58603 should be merged after this PR, and will resolve the issues that snipercup brought up.

File: data/json/ui/sidebar-legacy-compact.json
Comment (dseguin) [inline data/json/ui/sidebar-legacy-compact.json:137]:
```suggestion
      "bodygraph_hp_layout"
```

**Review by dseguin — APPROVED** [review]
Looks good! I was concerned about performance going from hardcoded methods to JSON conditions, so I profiled the before-and-after for the "classic" sidebar

##### before

before

##### after

after

Minor performance impact, I don't see it being a big issue. The only thing I would change is, whenever you display the body graph, you should use the layout that includes the hp graphs.

I didn't check all of the widgets, but I trust that folks will make let you know of any differences :)

**Pull request merged**
